### PR TITLE
Implement constraintsWithVisualFormat method on NSLayoutConstraint

### PR DESCRIPTION
**Pull request by BennyKJohnson**
This PR implements [constraintsWithVisualFormat](https://developer.apple.com/documentation/uikit/nslayoutconstraint/1526944-constraintswithvisualformat). I chose to place all the functionality in an auxiliary class due to its complexity. To the best of my knowledge it implements all the behaviour of the original implementation. My test suite which I developed alongside the code is located [here](https://github.com/BennyKJohnson/GSAutoLayoutVFLParser), this can be used as a guide to its expected behaviour, all the tests pass on both MacOS and Ubuntu with GNUStep.  I would be happy to contribute the test suite as well if there was motivation to support XCTest for this project. 

Let me know if I need to make any adjustments in terms of style etc @fredkiefer.

Thank you 

Screen Shot 2022-11-05 at 8 59 35 pm

### CODE REVIEW

File: Source/GSAutoLayoutVFLParser.m
Comment (gcasa) [inline Source/GSAutoLayoutVFLParser.m:1]: Same thing here.  Mainly the formatting is wrong.  Normally we go for the left brace on the next line and indented by two spaces instead of on the same line.   The algorithms here look correct, but I have not had a chance to test them in a running app.  I will do that tonight.   Also, we need to have the LGPL text on the top of both the source and the header files.

Comment (gcasa) [reply Source/GSAutoLayoutVFLParser.m:1]: BTW, I got your paperwork, thank you.

**Review by gcasa — CHANGES_REQUESTED** [review]
Great work, minor comments.  I am going to run a few tests and make some notes tonight.

**Comment by gcasa** [comment]
I have added @fredkiefer since he is the maintainer of GUI and reviews all changes.

**Comment by gcasa** [comment]
A decent rule of thumb on the formatting is to follow the suggestions I mentioned above and also to use emacs' C or C++ mode for editing ObjC.   This mode, generally, does the right thing.

File: Source/GSAutoLayoutVFLParser.h
Comment (BennyKJohnson) [reply Source/GSAutoLayoutVFLParser.h]: Hey @gcasa, thanks for your feedback. I've formatted the project using the GNU based style with clang format with some minor manual tweaks. Let me know if any of the formatting still needs work.

File: Source/GSAutoLayoutVFLParser.m
Comment (BennyKJohnson) [reply Source/GSAutoLayoutVFLParser.m:23]: Added the copyright notices

File: Source/GSAutoLayoutVFLParser.h
Comment (fredkiefer) [inline Source/GSAutoLayoutVFLParser.h:23]: Within the library itself we prefer to include only the headers that are actually needed, In this case Foundation, NSView and NSLayoutConstraint.

File: Source/GSAutoLayoutVFLParser.m
Comment (fredkiefer) [inline Source/GSAutoLayoutVFLParser.m:49]: In GNUstep we normally have a space after the colon.

I am not sure whether I like to see an exception being thrown from the init method. IN most cases this should result in the object not being cleaned up. I am not sure if even ARC would handle that correctly.

File: Source/GSAutoLayoutVFLParser.m
Comment (fredkiefer) [inline Source/GSAutoLayoutVFLParser.m:56]: All these are autoreleased object, they could go away when passed over the border of an autorelease pool. Most likely this is no issue here as this object will be short lived.

**Review by fredkiefer — COMMENTED** [review]
Sorry, it will take a bit of time for me to finish this review

File: Source/GSAutoLayoutVFLParser.m
Comment (BennyKJohnson) [reply Source/GSAutoLayoutVFLParser.m:49]: Regarding the potential memory leak, I did some research on this and your concern is correct however this is to be excepted when using NSException. From Apple's docs "The Cocoa frameworks are generally not exception-safe. The general pattern is that exceptions are reserved for programmer error only, and the program catching such an exception should quit soon afterwards." https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Exceptions/Articles/ExceptionsAndCocoaFrameworks.html#//apple_ref/doc/uid/TP40009045-SW1

File: Source/NSLayoutConstraint.m
Comment (BennyKJohnson) [reply Source/NSLayoutConstraint.m]: @fredkiefer I've address this in 1aff56cb9b355fab3d3a95f2473cafd97e69633e

File: Source/GSAutoLayoutVFLParser.h
Comment (BennyKJohnson) [reply Source/GSAutoLayoutVFLParser.h]: Addressed in 1aff56cb9b355fab3d3a95f2473cafd97e69633e

**Comment by BennyKJohnson** [comment]
@fredkiefer Thank you for your feedback, I pushed a commit addressing your feedback 1aff56cb9b355fab3d3a95f2473cafd97e69633e

File: Source/GSAutoLayoutVFLParser.m
Comment (fredkiefer) [inline Source/GSAutoLayoutVFLParser.m:530]: The `calloc` here is not strictly necessary. You could create this structure on the stack in `parsePredicateList` and pass it down. That way also the `free` won't be needed.
You don't need to change this, this is just an optimisation hint.

File: Source/GSAutoLayoutVFLParser.m
Comment (fredkiefer) [inline Source/GSAutoLayoutVFLParser.m:109]: I am not sure whether our minimal gcc version supports the object for loop. Maybe you would be better off to use the FOR_IN macro, see GSFastEnumeration.h

**Review by fredkiefer — COMMENTED** [review]
This code looks very nicely done. I will need to check your tests too. And of course it would be great to have the tests directly in libs-gui.

File: Source/GSAutoLayoutVFLParser.m
Comment (fredkiefer) [reply Source/GSAutoLayoutVFLParser.m]: Here I disagree with you. At least in the old times you would have returned `nil` here and as usual the caller has to handle that. Take a look at the documentation cited in the second answer here: https://stackoverflow.com/questions/6574938/why-not-throw-an-exception-if-super-init-returns-nil

**Comment by fredkiefer** [comment]
> This code looks very nicely done. I will need to check your tests too. And of course it would be great to have the tests directly in libs-gui.

I read the tests you wrote and should have done so earlier (Sorry again for being so slow). Very impressive test code. The next step will be to run your code with gcc to see what problems we get there. There used to be some integration pipeline for gui too but it is broken now. I really need to fix it, so much to do...

**Comment by BennyKJohnson** [comment]
Cheers @fredkiefer, I've pushed a new commit to address your feedback. Is there any documentation you can point me to that can guide me with setting up GCC for my environment?

**Review by fredkiefer — APPROVED** [review]
Looks great now! I hope to find time later this weekend to run this with gcc but I see nothing in the code that could cause trouble.

**Comment by BennyKJohnson** [comment]
@fredkiefer Any update on compiling with GCC?

> Looks great now! I hope to find time later this weekend to run this with gcc but I see nothing in the code that could cause trouble.

**Comment by fredkiefer** [comment]
> @fredkiefer Any update on compiling with GCC?

Sorry, I am a bit ill at the moment, nothing serious, just an annoying cold. But it keeps me from doing stuff for GNUstep in my spare time.

**Comment by fredkiefer** [comment]
I finally got around to test this on gcc. The good news is it worked, the bad we get a lot of warnings that should be addressed before merging this:

GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser initWithFormat:options:metrics:views:]':
GSAutoLayoutVFLParser.m:45:7: warning: suggest parentheses around assignment used as truth value [-Wparentheses]
   45 |   if (self = [super init])
      |       ^~~~
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser parse]':
GSAutoLayoutVFLParser.m:63:7: warning: 'GSAutoLayoutVFLParser' may not respond to '-failParseWithMessage:'
   63 |       [self failParseWithMessage: @"Cannot parse an empty string"];
      |       ^
GSAutoLayoutVFLParser.m:63:7: warning: (messages without a matching method signature will be assumed to return 'id' and accept '...' as arguments)
GSAutoLayoutVFLParser.m:66:3: warning: 'GSAutoLayoutVFLParser' may not respond to '-parseOrientation'
   66 |   [self parseOrientation];
      |   ^
GSAutoLayoutVFLParser.m:67:3: warning: 'GSAutoLayoutVFLParser' may not respond to '-parseLeadingSuperViewConnection'
   67 |   NSNumber *spacingConstant = [self parseLeadingSuperViewConnection];
      |   ^~~~~~~~
GSAutoLayoutVFLParser.m:72:7: warning: 'GSAutoLayoutVFLParser' may not respond to '-parseView'
   72 |       NSArray *viewConstraints = [self parseView];
      |       ^~~~~~~
GSAutoLayoutVFLParser.m:75:11: warning: 'GSAutoLayoutVFLParser' may not respond to '-addLeadingSuperviewConstraint:'
   75 |           [self addLeadingSuperviewConstraint: spacingConstant];
      |           ^
GSAutoLayoutVFLParser.m:82:29: warning: 'GSAutoLayoutVFLParser' may not respond to '-addViewSpacingConstraint:previousView:'
   82 |                             previousView: previousView];
      |                             ^~~~~~~~~~~~
GSAutoLayoutVFLParser.m:83:11: warning: 'GSAutoLayoutVFLParser' may not respond to '-addFormattingConstraints:'
   83 |           [self addFormattingConstraints: previousView];
      |           ^
GSAutoLayoutVFLParser.m:87:7: warning: 'GSAutoLayoutVFLParser' may not respond to '-parseConnection'
   87 |       spacingConstant = [self parseConnection];
      |       ^~~~~~~~~~~~~~~
GSAutoLayoutVFLParser.m:88:7: warning: passing argument 2 of 'scanString:intoString:' from incompatible pointer type [-Wincompatible-pointer-types]
   88 |       if ([_scanner scanString: @"|" intoString: nil])
      |       ^~
      |       |
      |       id
GSAutoLayoutVFLParser.m:88:7: note: expected 'struct NSString **' but argument is of type 'id'
GSAutoLayoutVFLParser.m:90:11: warning: 'GSAutoLayoutVFLParser' may not respond to '-addTrailingToSuperviewConstraint:'
   90 |           [self addTrailingToSuperviewConstraint: spacingConstant];
      |           ^
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser addFormattingConstraints:]':
GSAutoLayoutVFLParser.m:107:3: warning: 'GSAutoLayoutVFLParser' may not respond to '-assertHasValidFormatLayoutOptions'
  107 |   [self assertHasValidFormatLayoutOptions];
      |   ^
GSAutoLayoutVFLParser.m:109:3: warning: 'GSAutoLayoutVFLParser' may not respond to '-layoutAttributesForLayoutFormatOptions:'
  109 |   NSArray *attributes = [self layoutAttributesForLayoutFormatOptions:_options];
      |   ^~~~~~~
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser assertHasValidFormatLayoutOptions]':
GSAutoLayoutVFLParser.m:126:7: warning: 'GSAutoLayoutVFLParser' may not respond to '-isVerticalEdgeFormatLayoutOption:'
  126 |       [self isVerticalEdgeFormatLayoutOption: _options])
      |       ^
GSAutoLayoutVFLParser.m:129:34: warning: 'GSAutoLayoutVFLParser' may not respond to '-failParseWithMessage:'
  129 |                                  @"be used with a vertical layout"];
      |                                  ^
GSAutoLayoutVFLParser.m:132:12: warning: 'GSAutoLayoutVFLParser' may not respond to '-isVerticalEdgeFormatLayoutOption:'
  132 |            && ![self isVerticalEdgeFormatLayoutOption: _options])
      |            ^~
GSAutoLayoutVFLParser.m:135:34: warning: 'GSAutoLayoutVFLParser' may not respond to '-failParseWithMessage:'
  135 |                                  @"cannot be used with a horizontal layout"];
      |                                  ^
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser parseOrientation]':
GSAutoLayoutVFLParser.m:141:3: warning: passing argument 2 of 'scanString:intoString:' from incompatible pointer type [-Wincompatible-pointer-types]
  141 |   if ([_scanner scanString: @"V:" intoString: nil])
      |   ^~
      |   |
      |   id
GSAutoLayoutVFLParser.m:141:3: note: expected 'struct NSString **' but argument is of type 'id'
GSAutoLayoutVFLParser.m:147:7: warning: passing argument 2 of 'scanString:intoString:' from incompatible pointer type [-Wincompatible-pointer-types]
  147 |       [_scanner scanString: @"H:" intoString: nil];
      |       ^
      |       |
      |       id
GSAutoLayoutVFLParser.m:147:7: note: expected 'struct NSString **' but argument is of type 'id'
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser parseView]':
GSAutoLayoutVFLParser.m:153:3: warning: 'GSAutoLayoutVFLParser' may not respond to '-parseViewOpen'
  153 |   [self parseViewOpen];
      |   ^
GSAutoLayoutVFLParser.m:155:3: warning: 'GSAutoLayoutVFLParser' may not respond to '-parseViewName'
  155 |   _view = [self parseViewName];
      |   ^~~~~
GSAutoLayoutVFLParser.m:156:3: warning: 'GSAutoLayoutVFLParser' may not respond to '-parsePredicateList'
  156 |   NSArray *viewConstraints = [self parsePredicateList];
      |   ^~~~~~~
GSAutoLayoutVFLParser.m:157:3: warning: 'GSAutoLayoutVFLParser' may not respond to '-parseViewClose'
  157 |   [self parseViewClose];
      |   ^
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser parseViewName]':
GSAutoLayoutVFLParser.m:171:7: warning: 'GSAutoLayoutVFLParser' may not respond to '-failParseWithMessage:'
  171 |       [self failParseWithMessage: @"Failed to parse view name"];
      |       ^
GSAutoLayoutVFLParser.m:174:3: warning: 'GSAutoLayoutVFLParser' may not respond to '-isValidIdentifier:'
  174 |   if (![self isValidIdentifier: viewName])
      |   ^~
GSAutoLayoutVFLParser.m:178:17: warning: 'GSAutoLayoutVFLParser' may not respond to '-failParseWithMessage:'
  178 |                 @"and may only contain letters, numbers and underscores"];
      |                 ^
GSAutoLayoutVFLParser.m:181:3: warning: 'GSAutoLayoutVFLParser' may not respond to '-resolveViewWithIdentifier:'
  181 |   return [self resolveViewWithIdentifier: viewName];
      |   ^~~~~~
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser parseLeadingSuperViewConnection]':
GSAutoLayoutVFLParser.m:360:3: warning: passing argument 2 of 'scanString:intoString:' from incompatible pointer type [-Wincompatible-pointer-types]
  360 |   BOOL foundSuperview = [_scanner scanString: @"|" intoString: nil];
      |   ^~~~
      |   |
      |   id
GSAutoLayoutVFLParser.m:360:3: note: expected 'struct NSString **' but argument is of type 'id'
GSAutoLayoutVFLParser.m:366:3: warning: 'GSAutoLayoutVFLParser' may not respond to '-parseConnection'
  366 |   return [self parseConnection];
      |   ^~~~~~
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser parseConnection]':
GSAutoLayoutVFLParser.m:371:3: warning: passing argument 2 of 'scanString:intoString:' from incompatible pointer type [-Wincompatible-pointer-types]
  371 |   BOOL foundConnection = [_scanner scanString: @"-" intoString: nil];
      |   ^~~~
      |   |
      |   id
GSAutoLayoutVFLParser.m:371:3: note: expected 'struct NSString **' but argument is of type 'id'
GSAutoLayoutVFLParser.m:377:3: warning: 'GSAutoLayoutVFLParser' may not respond to '-parseSimplePredicate'
  377 |   NSNumber *simplePredicateValue = [self parseSimplePredicate];
      |   ^~~~~~~~
GSAutoLayoutVFLParser.m:378:3: warning: passing argument 2 of 'scanString:intoString:' from incompatible pointer type [-Wincompatible-pointer-types]
  378 |   BOOL endConnectionFound = [_scanner scanString: @"-" intoString: nil];
      |   ^~~~
      |   |
      |   id
GSAutoLayoutVFLParser.m:378:3: note: expected 'struct NSString **' but argument is of type 'id'
GSAutoLayoutVFLParser.m:382:7: warning: 'GSAutoLayoutVFLParser' may not respond to '-failParseWithMessage:'
  382 |       [self failParseWithMessage: @"A connection must end with a '-'"];
      |       ^
GSAutoLayoutVFLParser.m:386:7: warning: 'GSAutoLayoutVFLParser' may not respond to '-failParseWithMessage:'
  386 |       [self failParseWithMessage: @"Found invalid connection"];
      |       ^
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser parseSimplePredicate]':
GSAutoLayoutVFLParser.m:412:7: warning: 'GSAutoLayoutVFLParser' may not respond to '-isValidIdentifier:'
  412 |       if (![self isValidIdentifier: metricName])
      |       ^~
GSAutoLayoutVFLParser.m:417:21: warning: 'GSAutoLayoutVFLParser' may not respond to '-failParseWithMessage:'
  417 |                     @"numbers and underscores"];
      |                     ^
GSAutoLayoutVFLParser.m:420:7: warning: 'GSAutoLayoutVFLParser' may not respond to '-resolveMetricWithIdentifier:'
  420 |       return [self resolveMetricWithIdentifier: metricName];
      |       ^~~~~~
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser parsePredicateList]':
GSAutoLayoutVFLParser.m:426:3: warning: passing argument 2 of 'scanString:intoString:' from incompatible pointer type [-Wincompatible-pointer-types]
  426 |   BOOL startsWithPredicateList = [_scanner scanString: @"(" intoString: nil];
      |   ^~~~
      |   |
      |   id
GSAutoLayoutVFLParser.m:426:3: note: expected 'struct NSString **' but argument is of type 'id'
GSAutoLayoutVFLParser.m:437:7: warning: 'GSAutoLayoutVFLParser' may not respond to '-parseObjectOfPredicate:'
  437 |       [self parseObjectOfPredicate: &predicate];
      |       ^
GSAutoLayoutVFLParser.m:439:11: warning: 'GSAutoLayoutVFLParser' may not respond to '-createConstraintFromParsedPredicate:'
  439 |           addObject: [self createConstraintFromParsedPredicate: &predicate]];
      |           ^~~~~~~~~
GSAutoLayoutVFLParser.m:441:7: warning: passing argument 2 of 'scanString:intoString:' from incompatible pointer type [-Wincompatible-pointer-types]
  441 |       shouldParsePredicate = [_scanner scanString: @"," intoString: nil];
      |       ^~~~~~~~~~~~~~~~~~~~
      |       |
      |       id
GSAutoLayoutVFLParser.m:441:7: note: expected 'struct NSString **' but argument is of type 'id'
GSAutoLayoutVFLParser.m:444:3: warning: passing argument 2 of 'scanString:intoString:' from incompatible pointer type [-Wincompatible-pointer-types]
  444 |   if (![_scanner scanString: @")" intoString: nil])
      |   ^~
      |   |
      |   id
GSAutoLayoutVFLParser.m:444:3: note: expected 'struct NSString **' but argument is of type 'id'
GSAutoLayoutVFLParser.m:446:7: warning: 'GSAutoLayoutVFLParser' may not respond to '-failParseWithMessage:'
  446 |       [self failParseWithMessage: @"A predicate on a view must end with ')'"];
      |       ^
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser parseObjectOfPredicate:]':
GSAutoLayoutVFLParser.m:491:3: warning: 'GSAutoLayoutVFLParser' may not respond to '-parseRelation'
  491 |   NSLayoutRelation relation = [self parseRelation];
      |   ^~~~~~~~~~~~~~~~
GSAutoLayoutVFLParser.m:491:31: warning: initialization of 'NSLayoutRelation' {aka 'long int'} from 'id' makes integer from pointer without a cast [-Wint-conversion]
  491 |   NSLayoutRelation relation = [self parseRelation];
      |                               ^
GSAutoLayoutVFLParser.m:498:7: warning: 'GSAutoLayoutVFLParser' may not respond to '-parseIdentifier'
  498 |       NSString *identiferName = [self parseIdentifier];
      |       ^~~~~~~~
GSAutoLayoutVFLParser.m:499:7: warning: 'GSAutoLayoutVFLParser' may not respond to '-isValidIdentifier:'
  499 |       if (![self isValidIdentifier: identiferName])
      |       ^~
GSAutoLayoutVFLParser.m:504:21: warning: 'GSAutoLayoutVFLParser' may not respond to '-failParseWithMessage:'
  504 |                     @"contain letters, numbers and underscores"];
      |                     ^
GSAutoLayoutVFLParser.m:523:11: warning: 'GSAutoLayoutVFLParser' may not respond to '-failParseWithMessage:'
  523 |           [self failParseWithMessage: message];
      |           ^
GSAutoLayoutVFLParser.m:527:3: warning: 'GSAutoLayoutVFLParser' may not respond to '-parsePriority'
  527 |   NSNumber *priorityValue = [self parsePriority];
      |   ^~~~~~~~
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser parseRelation]':
GSAutoLayoutVFLParser.m:537:3: warning: passing argument 2 of 'scanString:intoString:' from incompatible pointer type [-Wincompatible-pointer-types]
  537 |   if ([_scanner scanString: @"==" intoString: nil])
      |   ^~
      |   |
      |   id
GSAutoLayoutVFLParser.m:537:3: note: expected 'struct NSString **' but argument is of type 'id'
GSAutoLayoutVFLParser.m:541:3: warning: passing argument 2 of 'scanString:intoString:' from incompatible pointer type [-Wincompatible-pointer-types]
  541 |   else if ([_scanner scanString: @">=" intoString: nil])
      |   ^~~~
      |   |
      |   id
GSAutoLayoutVFLParser.m:541:3: note: expected 'struct NSString **' but argument is of type 'id'
GSAutoLayoutVFLParser.m:545:3: warning: passing argument 2 of 'scanString:intoString:' from incompatible pointer type [-Wincompatible-pointer-types]
  545 |   else if ([_scanner scanString: @"<=" intoString: nil])
      |   ^~~~
      |   |
      |   id
GSAutoLayoutVFLParser.m:545:3: note: expected 'struct NSString **' but argument is of type 'id'
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser parsePriority]':
GSAutoLayoutVFLParser.m:561:28: warning: passing argument 2 of 'scanCharactersFromSet:intoString:' from incompatible pointer type [-Wincompatible-pointer-types]
  561 |                            intoString: nil];
      |                            ^~~~~~~~~~
      |                            |
      |                            id
GSAutoLayoutVFLParser.m:561:28: note: expected 'struct NSString **' but argument is of type 'id'
GSAutoLayoutVFLParser.m:567:3: warning: 'GSAutoLayoutVFLParser' may not respond to '-parseConstant'
  567 |   return [self parseConstant];
      |   ^~~~~~
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser resolveMetricWithIdentifier:]':
GSAutoLayoutVFLParser.m:575:7: warning: 'GSAutoLayoutVFLParser' may not respond to '-failParseWithMessage:'
  575 |       [self failParseWithMessage: @"Found metric not inside metric dictionary"];
      |       ^
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser resolveViewWithIdentifier:]':
GSAutoLayoutVFLParser.m:585:7: warning: 'GSAutoLayoutVFLParser' may not respond to '-failParseWithMessage:'
  585 |       [self failParseWithMessage: @"Found view not inside view dictionary"];
      |       ^
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser parseConstant]':
GSAutoLayoutVFLParser.m:599:3: warning: 'GSAutoLayoutVFLParser' may not respond to '-parseIdentifier'
  599 |   NSString *metricName = [self parseIdentifier];
      |   ^~~~~~~~
GSAutoLayoutVFLParser.m:600:3: warning: 'GSAutoLayoutVFLParser' may not respond to '-isValidIdentifier:'
  600 |   if (![self isValidIdentifier: metricName])
      |   ^~
GSAutoLayoutVFLParser.m:605:17: warning: 'GSAutoLayoutVFLParser' may not respond to '-failParseWithMessage:'
  605 |                 @"and underscores"];
      |                 ^
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser parseIdentifier]':
GSAutoLayoutVFLParser.m:621:7: warning: 'GSAutoLayoutVFLParser' may not respond to '-failParseWithMessage:'
  621 |       [self failParseWithMessage: @"Failed to find constant or metric"];
      |       ^
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser parseViewOpen]':
GSAutoLayoutVFLParser.m:632:50: warning: passing argument 2 of 'scanCharactersFromSet:intoString:' from incompatible pointer type [-Wincompatible-pointer-types]
  632 |                                                  intoString: nil];
      |                                                  ^~~~~~~~~~
      |                                                  |
      |                                                  id
GSAutoLayoutVFLParser.m:632:50: note: expected 'struct NSString **' but argument is of type 'id'
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser parseViewClose]':
GSAutoLayoutVFLParser.m:646:7: warning: passing argument 2 of 'scanCharactersFromSet:intoString:' from incompatible pointer type [-Wincompatible-pointer-types]
  646 |       [_scanner scanCharactersFromSet: closeViewIdentifier intoString: nil];
      |       ^
      |       |
      |       id
GSAutoLayoutVFLParser.m:646:7: note: expected 'struct NSString **' but argument is of type 'id'
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser isValidIdentifier:]':
GSAutoLayoutVFLParser.m:660:30: warning: passing argument 3 of 'regularExpressionWithPattern:options:error:' from incompatible pointer type [-Wincompatible-pointer-types]
  660 |                              error: nil];
      |                              ^~~~~
      |                              |
      |                              id
GSAutoLayoutVFLParser.m:660:30: note: expected 'struct NSError **' but argument is of type 'id'
GSAutoLayoutVFLParser.m: In function '-[GSAutoLayoutVFLParser layoutAttributesForLayoutFormatOptions:]':
GSAutoLayoutVFLParser.m:727:7: warning: 'GSAutoLayoutVFLParser' may not respond to '-failParseWithMessage:'
  727 |       [self failParseWithMessage:@"Unrecognized layout formatting option"];
      |       ^


Most of these come from the fact that gcc does not like the usage of methods that will be defined later in the same file. Here it is sufficient to either add a forward definition at the start of the file or rearrange methods. For the other issue you will have to use `NULL` instead of `nil`.  Both are easy to fix.

**Comment by BennyKJohnson** [comment]
@fredkiefer Thanks for compiling the code with GCC. I'm willing to work through the issues but forgive my lack of knowledge, I don't know how to compile the project using GCC. Could you please provide some guidance or point me to some documentation so I can replicate these errors on my environment.

**Comment by fredkiefer** [comment]
Compilation with gcc is fairly easy. Just install it and use it. What is complicated is to have one computer where you are able to switch between clang and gcc. I don't have any experience with such a setup. Most of us just use one or the other or have different machines for the different environments. The simplest solution will probably be to set up a virtual machine or Docker image with gcc installed and compile there.
But you don't need that. Just fix the issues gcc reported and I will try again and things should work. I promise to be a bit faster for the next round. My cold is almost over now.

**Comment by BennyKJohnson** [comment]
@fredkiefer I've updated the PR with fixes to the GCC warnings

**Pull request merged**
